File: ts-src/permissions/rules.ts
import type { ToolCategory } from "./checker.ts";

export interface PermissionRule {
	tool: string; // tool name, or "*"
	pattern?: string; // glob pattern (optional)
}

export interface PermissionConfig {
	allow: PermissionRule[];
	deny: PermissionRule[];
	ask: PermissionRule[];
}

// ── Tool name mapping ──────────────────────────────────────────────────
// Maps PascalCase display names to internal tool names
const TOOL_NAME_MAP: Record<string, string> = {
	read: "read_file",
	write: "write_file",
	edit: "edit_file",
	bash: "bash",
	glob: "glob",
	grep: "grep",
	webfetch: "web_fetch",
	askuser: "ask_user",
	savememory: "save_memory",
};

// Known internal tool names (already snake_case)
const KNOWN_TOOLS = new Set([
	"read_file",
	"write_file",
	"edit_file",
	"bash",
	"glob",
	"grep",
	"web_fetch",
	"ask_user",
	"save_memory",
]);

const CATEGORY_TOOLS: Record<ToolCategory, string[]> = {
	read: ["read_file", "glob", "grep", "ask_user"],
	write: ["write_file", "edit_file", "save_memory"],
	execute: ["bash"],
	network: ["web_fetch"],
};

const CATEGORY_NAMES = new Set(Object.keys(CATEGORY_TOOLS));

// Tools whose primary arg is a file path
const PATH_TOOLS = new Set(["read_file", "write_file", "edit_file", "glob", "grep"]);

// ── Glob cache ─────────────────────────────────────────────────────────
const globCache = new Map<string, InstanceType<typeof Bun.Glob>>();

function getGlob(pattern: string): InstanceType<typeof Bun.Glob> {
	let g = globCache.get(pattern);
	if (!g) {
		g = new Bun.Glob(pattern);
		globCache.set(pattern, g);
	}
	return g;
}

function globMatch(pattern: string, str: string): boolean {
	return getGlob(pattern).match(str);
}

// ── Helpers ────────────────────────────────────────────────────────────
function basename(filePath: string): string {
	const parts = filePath.split("/");
	return parts[parts.length - 1] ?? filePath;
}

/**
 * Match a command pattern against a command string.
 * "rm *" matches any command starting with "rm " (prefix match).
 * "rm" matches exactly "rm".
 * "git status" matches exactly "git status".
 */
function matchCommand(pattern: string, command: string): boolean {
	// Pattern ending with " *" → prefix match (everything starting with that prefix)
	if (pattern.endsWith(" *")) {
		const prefix = pattern.slice(0, -1); // keep the trailing space: "rm "
		return command.startsWith(prefix) || command === pattern.slice(0, -2);
	}
	// Exact match
	return command === pattern;
}

function extractHostname(url: string): string | null {
	try {
		return new URL(url).hostname;
	} catch {
		return null;
	}
}

function resolveToolName(name: string): string | null {
	const lower = name.toLowerCase();
	// Check if it's already a known internal name
	if (KNOWN_TOOLS.has(lower)) return lower;
	// Check display name map (case-insensitive, no separators)
	const normalized = lower.replace(/[_-]/g, "");
	if (TOOL_NAME_MAP[normalized]) return TOOL_NAME_MAP[normalized];
	return null;
}

// ── Parse ──────────────────────────────────────────────────────────────

/**
 * Parse a rule string like "Write(src/**)" into PermissionRule(s).
 * Returns null for invalid strings.
 * Returns an array for category names (expands to all tools in category).
 */
export function parseRule(raw: string): PermissionRule | PermissionRule[] | null {
	const trimmed = raw.trim();
	if (!trimmed) return null;

	// Wildcard
	if (trimmed === "*") return { tool: "*" };

	// Extract name and optional pattern
	let name: string;
	let pattern: string | undefined;

	const parenOpen = trimmed.indexOf("(");
	if (parenOpen !== -1) {
		if (!trimmed.endsWith(")")) return null; // unclosed paren
		name = trimmed.slice(0, parenOpen);
		pattern = trimmed.slice(parenOpen + 1, -1);
	} else {
		name = trimmed;
	}

	const lowerName = name.toLowerCase();

	// Lowercase names that match a category are treated as categories
	// (e.g., "write" → category, "Write" → tool name write_file)
	if (name === lowerName && CATEGORY_NAMES.has(lowerName)) {
		const tools = CATEGORY_TOOLS[lowerName as ToolCategory];
		return tools.map((tool) => (pattern ? { tool, pattern } : { tool }));
	}

	// Resolve as specific tool name
	const toolName = resolveToolName(name);
	if (toolName) {
		return pattern ? { tool: toolName, pattern } : { tool: toolName };
	}

	// Unknown name — return null
	return null;
}

// ── Match ──────────────────────────────────────────────────────────────

/** Does this rule match the given tool call? */
export function matchRule(rule: PermissionRule, toolName: string, args?: Record<string, unknown>): boolean {
	// Tool name match
	if (rule.tool !== "*" && rule.tool !== toolName) return false;

	// No pattern = bare name match (any invocation)
	if (!rule.pattern) return true;

	// Pattern requires args to match against
	if (!args) return false;

	// File path tools: match against path arg (full path + basename)
	if (PATH_TOOLS.has(toolName) && args.path) {
		const filePath = String(args.path);
		return globMatch(rule.pattern, filePath) || globMatch(rule.pattern, basename(filePath));
	}

	// Bash: match against command string using prefix matching
	// Glob * doesn't cross / which is wrong for commands, so use startsWith for "cmd *" patterns
	if (toolName === "bash" && args.command) {
		const cmd = String(args.command);
		return matchCommand(rule.pattern, cmd);
	}

	// Web fetch: match against hostname
	if (toolName === "web_fetch" && args.url) {
		const hostname = extractHostname(String(args.url));
		if (hostname) return globMatch(rule.pattern, hostname);
	}

	return false;
}

// ── Evaluate ───────────────────────────────────────────────────────────

/**
 * Evaluate rules against a tool call.
 * Priority: deny > ask > allow.
 * Returns null if no rules match.
 */
export function evaluateRules(
	config: PermissionConfig,
	toolName: string,
	args?: Record<string, unknown>,
): "allow" | "deny" | "ask" | null {
	const denyMatch = config.deny.some((r) => matchRule(r, toolName, args));
	if (denyMatch) return "deny";

	const askMatch = config.ask.some((r) => matchRule(r, toolName, args));
	if (askMatch) return "ask";

	const allowMatch = config.allow.some((r) => matchRule(r, toolName, args));
	if (allowMatch) return "allow";

	return null;
}

// ── Merge ──────────────────────────────────────────────────────────────

/**
 * Merge permission configs from multiple layers (global → local → overrides).
 * More specific layers (later args) can override less specific layers:
 * - If a later layer explicitly allows something a prior layer denied (same tool+pattern),
 *   the deny is removed so the allow takes effect.
 * - If a later layer explicitly denies something a prior layer allowed,
 *   the deny takes precedence (deny wins within merged result).
 */
export function mergeConfigs(...configs: PermissionConfig[]): PermissionConfig {
	if (configs.length === 0) return { allow: [], deny: [], ask: [] };
	if (configs.length === 1) {
		const c = configs[0] as PermissionConfig;
		return { allow: [...c.allow], deny: [...c.deny], ask: [...c.ask] };
	}

	// Track accumulated rules, allowing later layers to override
	const mergedDeny: PermissionRule[] = [];
	const mergedAllow: PermissionRule[] = [];
	const mergedAsk: PermissionRule[] = [];

	for (const config of configs) {
		// Later-layer allows can remove prior-layer denies (same tool+pattern override)
		for (const allowRule of config.allow) {
			// Remove matching denies from earlier layers
			const idx = mergedDeny.findIndex((d) => d.tool === allowRule.tool && d.pattern === allowRule.pattern);
			if (idx !== -1) {
				mergedDeny.splice(idx, 1);
			}
			mergedAllow.push(allowRule);
		}

		// Later-layer denies always stack
		for (const denyRule of config.deny) {
			mergedDeny.push(denyRule);
		}

		// Ask rules stack
		for (const askRule of config.ask) {
			mergedAsk.push(askRule);
		}
	}

	return { allow: mergedAllow, deny: mergedDeny, ask: mergedAsk };
}

File: src/permissions/rules.rs
//! Rule parsing and evaluation. Mirrors `ts-src/permissions/rules.ts`.

use std::collections::HashMap;

use globset::Glob;
use once_cell::sync::Lazy;
use serde_json::Value;
use url::Url;

use super::checker::ToolCategory;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PermissionRule {
    pub tool: String,
    pub pattern: Option<String>,
}

#[derive(Debug, Clone, Default)]
pub struct PermissionConfig {
    pub allow: Vec<PermissionRule>,
    pub deny: Vec<PermissionRule>,
    pub ask: Vec<PermissionRule>,
}

static TOOL_NAME_MAP: Lazy<HashMap<&'static str, &'static str>> = Lazy::new(|| {
    let mut m = HashMap::new();
    m.insert("read", "read_file");
    m.insert("write", "write_file");
    m.insert("edit", "edit_file");
    m.insert("bash", "bash");
    m.insert("glob", "glob");
    m.insert("grep", "grep");
    m.insert("webfetch", "web_fetch");
    m.insert("askuser", "ask_user");
    m.insert("savememory", "save_memory");
    m
});

static KNOWN_TOOLS: Lazy<Vec<&'static str>> = Lazy::new(|| {
    vec![
        "read_file",
        "write_file",
        "edit_file",
        "bash",
        "glob",
        "grep",
        "web_fetch",
        "ask_user",
        "save_memory",
    ]
});

static CATEGORY_TOOLS: Lazy<HashMap<&'static str, Vec<&'static str>>> = Lazy::new(|| {
    let mut m = HashMap::new();
    m.insert("read", vec!["read_file", "glob", "grep", "ask_user"]);
    m.insert("write", vec!["write_file", "edit_file", "save_memory"]);
    m.insert("execute", vec!["bash"]);
    m.insert("network", vec!["web_fetch"]);
    m
});

const PATH_TOOLS: &[&str] = &["read_file", "write_file", "edit_file", "glob", "grep"];

fn glob_match(pattern: &str, target: &str) -> bool {
    Glob::new(pattern)
        .map(|g| g.compile_matcher().is_match(target))
        .unwrap_or(false)
}

fn basename(file_path: &str) -> &str {
    file_path.rsplit('/').next().unwrap_or(file_path)
}

/// Matches a command pattern against a command string.
/// "rm *" → prefix match starting with "rm "; "rm" → exact match.
fn match_command(pattern: &str, command: &str) -> bool {
    if let Some(prefix) = pattern.strip_suffix(" *") {
        return command.starts_with(&format!("{prefix} ")) || command == prefix;
    }
    command == pattern
}

fn extract_hostname(url: &str) -> Option<String> {
    Url::parse(url)
        .ok()
        .and_then(|u| u.host_str().map(String::from))
}

fn resolve_tool_name(name: &str) -> Option<&'static str> {
    let lower = name.to_lowercase();
    if let Some(known) = KNOWN_TOOLS.iter().find(|t| **t == lower) {
        return Some(known);
    }
    let normalized: String = lower.chars().filter(|c| *c != '_' && *c != '-').collect();
    TOOL_NAME_MAP.get(normalized.as_str()).copied()
}

/// Parse a rule string like `"Write(src/**)"`. Returns one rule, a list (for
/// category names that expand to all tools), or `None` for unknown names.
pub enum ParsedRule {
    One(PermissionRule),
    Many(Vec<PermissionRule>),
}

pub fn parse_rule(raw: &str) -> Option<ParsedRule> {
    let trimmed = raw.trim();
    if trimmed.is_empty() {
        return None;
    }
    if trimmed == "*" {
        return Some(ParsedRule::One(PermissionRule {
            tool: "*".to_string(),
            pattern: None,
        }));
    }

    let (name, pattern) = if let Some(paren_idx) = trimmed.find('(') {
        if !trimmed.ends_with(')') {
            return None;
        }
        let n = &trimmed[..paren_idx];
        let p = &trimmed[paren_idx + 1..trimmed.len() - 1];
        (n.to_string(), Some(p.to_string()))
    } else {
        (trimmed.to_string(), None)
    };

    let lower_name = name.to_lowercase();

    // Lowercase + category → expand to all tools in category
    if name == lower_name && CATEGORY_TOOLS.contains_key(lower_name.as_str()) {
        let tools = &CATEGORY_TOOLS[lower_name.as_str()];
        let rules: Vec<PermissionRule> = tools
            .iter()
            .map(|t| PermissionRule {
                tool: t.to_string(),
                pattern: pattern.clone(),
            })
            .collect();
        return Some(ParsedRule::Many(rules));
    }

    // Resolve specific tool
    let tool_name = resolve_tool_name(&name)?;
    Some(ParsedRule::One(PermissionRule {
        tool: tool_name.to_string(),
        pattern,
    }))
}

pub fn match_rule(rule: &PermissionRule, tool_name: &str, args: Option<&Value>) -> bool {
    if rule.tool != "*" && rule.tool != tool_name {
        return false;
    }
    let pattern = match &rule.pattern {
        Some(p) => p,
        None => return true,
    };
    let args = match args {
        Some(a) => a,
        None => return false,
    };

    if PATH_TOOLS.contains(&tool_name) {
        if let Some(path) = args.get("path").and_then(|v| v.as_str()) {
            return glob_match(pattern, path) || glob_match(pattern, basename(path));
        }
    }

    if tool_name == "bash" {
        if let Some(cmd) = args.get("command").and_then(|v| v.as_str()) {
            return match_command(pattern, cmd);
        }
    }

    if tool_name == "web_fetch" {
        if let Some(url) = args.get("url").and_then(|v| v.as_str()) {
            if let Some(host) = extract_hostname(url) {
                return glob_match(pattern, &host);
            }
        }
    }

    false
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RuleDecision {
    Allow,
    Deny,
    Ask,
}

pub fn evaluate_rules(
    config: &PermissionConfig,
    tool_name: &str,
    args: Option<&Value>,
) -> Option<RuleDecision> {
    if config.deny.iter().any(|r| match_rule(r, tool_name, args)) {
        return Some(RuleDecision::Deny);
    }
    if config.ask.iter().any(|r| match_rule(r, tool_name, args)) {
        return Some(RuleDecision::Ask);
    }
    if config.allow.iter().any(|r| match_rule(r, tool_name, args)) {
        return Some(RuleDecision::Allow);
    }
    None
}

pub fn merge_configs(configs: &[PermissionConfig]) -> PermissionConfig {
    if configs.is_empty() {
        return PermissionConfig::default();
    }
    if configs.len() == 1 {
        return configs[0].clone();
    }

    let mut merged_deny: Vec<PermissionRule> = Vec::new();
    let mut merged_allow: Vec<PermissionRule> = Vec::new();
    let mut merged_ask: Vec<PermissionRule> = Vec::new();

    for config in configs {
        // Allow in later layer removes matching deny from earlier layers
        for allow in &config.allow {
            if let Some(idx) = merged_deny
                .iter()
                .position(|d| d.tool == allow.tool && d.pattern == allow.pattern)
            {
                merged_deny.remove(idx);
            }
            merged_allow.push(allow.clone());
        }
        for deny in &config.deny {
            merged_deny.push(deny.clone());
        }
        for ask in &config.ask {
            merged_ask.push(ask.clone());
        }
    }

    PermissionConfig {
        allow: merged_allow,
        deny: merged_deny,
        ask: merged_ask,
    }
}

/// Stub so other modules can refer to category via this name (rebound below).
pub fn _category_tools() -> &'static HashMap<&'static str, Vec<&'static str>> {
    &CATEGORY_TOOLS
}

impl ToolCategory {
    pub fn name(&self) -> &'static str {
        match self {
            Self::Read => "read",
            Self::Write => "write",
            Self::Execute => "execute",
            Self::Network => "network",
        }
    }
}
